Add composer

johnstevenson opened the pull request:
<!-- Provide a general summary of your changes in the Title above, prefixed with (packageName) -->

## Description
This adds the [composer](https://chocolatey.org/packages/composer) package to the coreteampackages repo

## Motivation and Context
Apart from the obvious benefits of automated updates, this also gives the package some authority. This will be beneficial because the package requires exemption from verification testing, due to the current limitations of the test environment.

## How Has this Been Tested?
Tested locally and on Appveyor: https://ci.appveyor.com/project/johnstevenson/chocolatey-composer-setup/builds/29362407

## Screenshot (if appropriate, usually isn't needed):

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Migrated package (a package has been migrated from another repository)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this repository.
- [x] My change requires a change to documentation (this usually means the notes in the description of a package).
- [x] I have updated the documentation accordingly (this usually means the notes in the description of a package).
- [x] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey/chocolatey-coreteampackages/blob/master/CONTRIBUTING.md)
- [ ] The added/modified package passed install/uninstall in the chocolatey test environment.
- [x] The changes only affect a single package (not including meta package).

<!-- The following section can be removed if the package has not been migrated from another location -->
## Original Location
- [composer/windows-setup](https://github.com/composer/windows-setup)
- [Open Issues](https://github.com/composer/windows-setup/issues) Issues are generally related to the configuration of PHP (which is a dependency of this package so it will already be configured correctly)
- [x] No PR (some of the code may be kept here for testing local pre-release builds)
- [x] The [migration guidelines](https://github.com/chocolatey/chocolatey-coreteampackages/wiki/Package-migration-process) have been followed

johnstevenson commented:
Ha, ha. Welcome to the world of pain resulting from trying to install PHP on Appveyor!

In this instance it is failing to install because:
```
Unable to resolve dependency 'chocolatey-core.extension (= 1.3.3)'.
```
I'm sure you guys will have a better idea than me why this is happening, but there is always seems to be some issue.

This issue also happens with the latest PHP version (7.4.0), using the `--force` param. See: https://ci.appveyor.com/project/johnstevenson/chocolatey-composer-setup/builds/29366903

Without using the `--force` param is fails due to:
```
Failures
 - kb2999226 (exited 1058) - Error while running 'C:\ProgramData\chocolatey\lib\KB2999226\tools\chocolateyinstall.ps1'.
 See log for details.
 - kb2919442 (exited 1058) - Error while running 'C:\ProgramData\chocolatey\lib\KB2919442\tools\ChocolateyInstall.ps1'.
 See log for details.
 - kb3035131 (exited 1058) - Error while running 'C:\ProgramData\chocolatey\lib\KB3035131\Tools\ChocolateyInstall.ps1'.
 See log for details.
```
See: https://ci.appveyor.com/project/johnstevenson/chocolatey-composer-setup/builds/29344897
I guess this is something to do with the Windows Update Service?


The only reliable way that I have found to install PHP on Appveyor is to ignore its dependencies
(`cinst -i -y php`), which is what we do for Windows testing on some Composer repos:
https://github.com/composer/composer/blob/master/appveyor.yml
https://github.com/composer/xdebug-handler/blob/master/appveyor.yml

johnstevenson commented:
I guess the error was to do with this dependencies issue: https://status.chocolatey.org/issues/2019-11-29-slow-and-intermittent-connection-issues/

johnstevenson commented:
I updated the composer package and force-pushed this to see if we still got errors (which we apparently don't).

johnstevenson commented:
@AdmiringWorm Do I need to do anything else here?

I wasn't sure about how to add the logo into the readme - is that something that is done by yourselves?

AdmiringWorm commented:
@johnstevenson you don't need to add the icon yourself to the nuspec/readme. It will get automatically updated (as long as it finds something to update) when the package gets updated during our update checks runs.

I have not yet had a chance to have a look at this PR, so I can not yet say if there are anything you will need to change or not.
I hope to be able to review this PR later this week.

AdmiringWorm commented on automatic/composer/ReadMe.md:
```diff
@@ -0,0 +1,21 @@
+
+Composer Setup downloads and installs the latest version of Composer, the PHP Dependency Manager, so you can use it easily from the command line: just type `composer` from any location to use it.
+
+Note: The version number refers to the installer and not to Composer, which you can update by running `composer self-update` from your terminal.
```
Please move this line to a `## Notes` section in the description.

johnstevenson replied:
Done

AdmiringWorm commented on automatic/composer/ReadMe.md:
```diff
@@ -0,0 +1,21 @@
+
+## Package Specifics
+
+This package has a dependency on the Chocolatey PHP package. If this is not found, the latest version will be downloaded and installed first.
```
This should be under a `## Notes` section header.

johnstevenson replied:
Done

AdmiringWorm commented on automatic/composer/ReadMe.md:
```diff
@@ -0,0 +1,21 @@
+
+This package has a dependency on the Chocolatey PHP package. If this is not found, the latest version will be downloaded and installed first.
+
+If you encounter any problems with the installation, you can run it interactively using the `--notsilent` option.
```
Same as earlier

johnstevenson replied:
Done

AdmiringWorm commented on automatic/composer/ReadMe.md:
```diff
@@ -0,0 +1,21 @@
+
+## Advanced Usage
+
+The following installer arguments can be set. They are mainly intended for CI usage:
```
Would it be possible to implement these as package parameters?
This is preferable instead of using raw installer arguments.

johnstevenson replied:
Implemented as package parameters

AdmiringWorm commented on automatic/composer/ReadMe.md:
```diff
@@ -0,0 +1,21 @@
+# <img src="https://rawcdn.githack.com/composer/windows-setup/80f687d4ec2c5a199770c131525fd08ff93d0366/chocolatey/composer-logo.png" width="48" height="48"/> [composer](https://chocolatey.org/packages/composer)
+
+Composer Setup downloads and installs the latest version of Composer, the PHP Dependency Manager, so you can use it easily from the command line: just type `composer` from any location to use it.
+
```
Please add a `## Features` section in this description.
This can just list the different features that are available for this software.

johnstevenson replied:
Done

AdmiringWorm commented on automatic/composer/legal/VERIFICATION.txt:
```diff
@@ -0,0 +1,21 @@
+  - Use powershell function 'Get-FileHash'
+  - Use Chocolatey utility 'checksum.exe'
+
+  checksum32: 80422FEF0EA310C0F565D6028F34010EEEBC96A4EDEC3175156D85F4F0EEEF10
```
Also add the `checksumType: sha256` here.

johnstevenson replied:
Done

AdmiringWorm commented on automatic/composer/tools/chocolateyinstall.ps1:
```diff
@@ -0,0 +1,30 @@
+  checksumType = 'sha256'
+  silentArgs   = '/VERYSILENT /SUPPRESSMSGBOXES /NORESTART'
+  softwareName = 'composer*'
+}
```
missing `validExitCodes` argument.
At a minimum, this should be `validExitCodel = @(0)`

johnstevenson replied:
Done

AdmiringWorm commented on automatic/composer/tools/chocolateyinstall.ps1:
```diff
@@ -0,0 +1,30 @@
+}
+catch {
+
+  if ($env:ChocolateyExitCode -eq '1') {
```
Is it guaranteed that the exit code will be `1` only when the php settings are incorrect?

Not saying that you need to change anything here, but just want to know.

johnstevenson replied:
Yes, the exit code is always 1 when the installer aborts due to incorrect settings.

AdmiringWorm commented on automatic/composer/update.ps1:
```diff
@@ -0,0 +1,37 @@
+
+    ".\legal\VERIFICATION.txt"      = @{
+      "(?i)(\s+x32:).*"     = "`${1} $($Latest.URL32)"
+      "(?i)(checksum32:).*" = "`${1} $($Latest.Checksum32)"
```
Should also be able to replace the checksum type here (when added to the `VERIFICATION.txt` file).

johnstevenson replied:
Done

AdmiringWorm commented:
I forgot to mention, please also add yourself to the [CODEOWNERS](https://github.com/chocolatey-community/chocolatey-coreteampackages/blob/cb948ba3cb4b0ed252b4d88e25b8f9f89661c6cf/.github/CODEOWNERS) file.
This will let us know that you should be requested a review of when a PR is opened for this package in the future (and allows us to keep track of who the maintainers are for each package).

majkinetor commented:
I think appropriate location for this package is already good - Its part of the authors repo and chocolatey release should be part of its own automated build rather then what we have here. I don not have any problem in hosting it here, but I don't see any beneif either. Given that there i no feedback, I am adding pending closure tag. If you feel differently, please let us know.


> Unable to resolve dependency 'chocolatey-core.extension (= 1.3.3)'.

This is dependency on php itself that you depend on. Not sure why it says this, nor why the php package itself uses it as I don't see any core function used during quick glance - it was probably used before chocoaltey supported `Get-PackageParameters` nateivelly. If you still have a problem with it, please report issue on php package.

johnstevenson commented:
@majkinetor Apologies, I've got a bit behind. I'll respond as soon as I can if that is okay.

majkinetor commented:
Sure m8.

johnstevenson commented:
@majkinetor 

>     Unable to resolve dependency 'chocolatey-core.extension (= 1.3.3)'.

I haven't seen this message myself.

> This is dependency on php itself that you depend on. Not sure why it says this, nor why the php package itself uses it as I don't see any core function used during quick glance - it was probably used before chocoaltey supported Get-PackageParameters nateivelly. If you still have a problem with it, please report issue on php package.

So what am I supposed to do here? This package depends on `php`, which itself has a dependency on `chocolatey-core.extension` because it uses Get-PackageParameters. Does the php package not need this now?

Also, if `composer` is to use Get-PackageParameters, as requested by @AdmiringWorm in his code review, do I need to include this dependency myself?

Any advice would be appreciated because the documentation at https://chocolatey.org/docs/helpers-get-package-parameters completely baffles me. Thanks.

majkinetor commented:
You should concentrate on fix documentation as requested by @AdmiringWorm 

> Also, if composer is to use Get-PackageParameters, as requested by @AdmiringWorm in his code review, do I need to include this dependency myself?

No you don't as chocolatey now supports it nativelly.

johnstevenson commented:
@majkinetor Thanks. 

> You should concentrate on fix documentation as requested by @AdmiringWorm

That's what I'm doing.

johnstevenson commented:
@AdmiringWorm Many apologies for the delay in addressing your review (it's been a strange old time recently). I've fixed:

- formated the readme into Features, Package Parameters and Notes section
- changed the installargs to package parameters
- changed au_SearchReplace to update `checksum type` in verification.txt and `checksumType` in chocolateyInstall.ps1
- added myself to the CODEOWNERS file.

I'm still unclear what to do about the icon and its references.

majkinetor commented:
As soon as chocolaty user is added by gallery admins this can be merged. 
Thx for your work @johnstevenson .

/cc @AdmiringWorm @pauby @gep13

gep13 commented:
@majkinetor chocolatey user has been added as a maintainer on this package.
